File: apps/android/app/src/test/kotlin/com/loonext/android/features/inbox/InboxFiltersTest.kt
package com.loonext.android.features.inbox

import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #548 — which dimensions the inbox is arranged by, and that this phone agrees
 * with the laptop about it.
 *
 * Two halves. The behaviour tests assert the rule; the parity test reads
 * `packages/shared/src/inbox-filters.ts`, because this is a hand-port and nothing
 * about Kotlin says the original stayed put.
 *
 * The bug this file exists for: the predicate used to live here as
 * `hasFilterChips` and it excluded the status segment, so the sheet's Reset — drawn
 * directly above a Status section — was a control that gave a haptic and did
 * nothing.
 */
class InboxFiltersTest {

    private fun home(
        segment: String? = null,
        assignedToMe: Boolean = false,
        assigneeUserId: String? = null,
        tagId: String? = null,
        unreadOnly: Boolean = false,
        spamOnly: Boolean = false,
        snoozedOnly: Boolean = false,
        awaitingOnly: Boolean = false,
    ) = InboxFilterState(
        segment = segment,
        assignedToMe = assignedToMe,
        assigneeUserId = assigneeUserId,
        tagId = tagId,
        unreadOnly = unreadOnly,
        spamOnly = spamOnly,
        snoozedOnly = snoozedOnly,
        awaitingOnly = awaitingOnly,
    )

    @Test
    fun `nothing is filtered on the home view`() {
        assertFalse(isInboxFiltered(home()))
        assertEquals(emptyList<InboxFilterDimension>(), activeInboxFilters(home()))
    }

    @Test
    fun `the status segment counts, which is the whole bug`() {
        // Select Closed, press Reset, get a haptic and nothing else. That was #548.
        assertTrue(isInboxFiltered(home(segment = "closed")))
        assertEquals(
            listOf(InboxFilterDimension.SEGMENT),
            activeInboxFilters(home(segment = "closed")),
        )
    }

    @Test
    fun `Mine counts as the segment moving, with no separate assignee`() {
        val mine = home(assignedToMe = true)
        assertTrue(isInboxFiltered(mine))
        assertEquals(listOf(InboxFilterDimension.SEGMENT), activeInboxFilters(mine))
    }

    @Test
    fun `Mine subsumes a named assignee rather than counting both`() {
        // The request sends the viewer's own id and drops the named one, and the
        // sheet hides the assignee section while Mine is lit — so counting it is
        // how an empty Mine tab blamed a filter with nothing to un-set.
        assertEquals(
            listOf(InboxFilterDimension.SEGMENT),
            activeInboxFilters(home(assignedToMe = true, assigneeUserId = "u-2")),
        )
    }

    @Test
    fun `each chip counts on its own`() {
        assertEquals(
            listOf(InboxFilterDimension.ASSIGNEE),
            activeInboxFilters(home(assigneeUserId = "u-2")),
        )
        assertEquals(listOf(InboxFilterDimension.TAG), activeInboxFilters(home(tagId = "t-1")))
        assertEquals(
            listOf(InboxFilterDimension.UNREAD),
            activeInboxFilters(home(unreadOnly = true)),
        )
        assertEquals(listOf(InboxFilterDimension.SPAM), activeInboxFilters(home(spamOnly = true)))
        assertEquals(
            listOf(InboxFilterDimension.SNOOZED),
            activeInboxFilters(home(snoozedOnly = true)),
        )
        assertEquals(
            listOf(InboxFilterDimension.AWAITING),
            activeInboxFilters(home(awaitingOnly = true)),
        )
    }

    @Test
    fun `everything at once, in the declared order`() {
        assertEquals(
            listOf(
                InboxFilterDimension.SEGMENT,
                InboxFilterDimension.ASSIGNEE,
                InboxFilterDimension.TAG,
                InboxFilterDimension.UNREAD,
                InboxFilterDimension.SPAM,
                InboxFilterDimension.SNOOZED,
                InboxFilterDimension.AWAITING,
            ),
            activeInboxFilters(
                home(
                    segment = "all",
                    assigneeUserId = "u-2",
                    tagId = "t-1",
                    unreadOnly = true,
                    spamOnly = true,
                    snoozedOnly = true,
                    awaitingOnly = true,
                ),
            ),
        )
    }

    @Test
    fun `the segment alone is not a SECONDARY filter`() {
        // So the empty state keeps its better per-tab sentence: "No closed
        // conversations" beats "Nothing matches these filters" for somebody who
        // selected one tab and nothing else.
        assertFalse(hasSecondaryInboxFilters(home(segment = "closed")))
        assertFalse(hasSecondaryInboxFilters(home(assignedToMe = true)))
        assertTrue(hasSecondaryInboxFilters(home(segment = "closed", tagId = "t-1")))
    }

    @Test
    fun `secondary filters always imply something is filtered`() {
        // The property that makes one list better than two predicates.
        val states = listOf(
            home(),
            home(segment = "closed"),
            home(unreadOnly = true),
            home(assignedToMe = true, assigneeUserId = "u-2"),
            home(segment = "all", tagId = "t-1", spamOnly = true),
        )
        for (state in states) {
            if (hasSecondaryInboxFilters(state)) assertTrue(isInboxFiltered(state))
        }
    }

    // ---------------------------------------------------- against the original

    private fun repoFile(relative: String): String {
        var dir: File? = File("").absoluteFile
        while (dir != null) {
            val candidate = File(dir, relative)
            if (candidate.exists()) return candidate.readText()
            dir = dir.parentFile
        }
        throw AssertionError("$relative not found from ${File("").absolutePath}")
    }

    /**
     * The same seven dimensions, in the same order, as the shared module.
     *
     * Order is asserted rather than membership alone: [activeInboxFilters] returns
     * a list, and a client rendering "arranged by" chips from it would show them in
     * a different order from the laptop if the two drifted.
     */
    @Test
    fun `the dimensions match the shared module, in order`() {
        val shared = repoFile("packages/shared/src/inbox-filters.ts")
        val declared = Regex("""INBOX_FILTER_DIMENSIONS = \[([^\]]+)\]""")
            .find(shared)
            ?.groupValues
            ?.get(1)
            ?: throw AssertionError("INBOX_FILTER_DIMENSIONS is no longer an array literal")
        val names = Regex(""""([a-z_]+)"""").findAll(declared)
            .map { it.groupValues[1].uppercase() }
            .toList()
        assertEquals(names, InboxFilterDimension.entries.map { it.name })
    }

    /**
     * And that the shared module still masks the assignee under Mine.
     *
     * A grep rather than a second implementation: this is the one rule in the file
     * whose absence would be invisible in a boolean, because both answers are
     * "filtered" — only the LIST differs, and only the empty-state copy reads it.
     */
    @Test
    fun `the shared module still subsumes a named assignee under Mine`() {
        val shared = repoFile("packages/shared/src/inbox-filters.ts")
        assertTrue(
            "the Mine-subsumes-assignee guard has gone from the shared module",
            shared.contains("!state.assignedToMe && state.assigneeUserId !== null"),
        )
    }
}

File: packages/shared/src/inbox-filters.test.ts
import { describe, expect, it } from "vitest";

import {
  activeInboxFilters,
  hasSecondaryInboxFilters,
  isInboxFiltered,
  type InboxFilterState,
} from "./inbox-filters";

/** Nothing selected: the home view on every client. */
function home(overrides: Partial<InboxFilterState> = {}): InboxFilterState {
  return {
    segment: null,
    assignedToMe: false,
    assigneeUserId: null,
    tagId: null,
    unreadOnly: false,
    spamOnly: false,
    snoozedOnly: false,
    awaitingOnly: false,
    ...overrides,
  };
}

describe("#548 isInboxFiltered", () => {
  it("says nothing is filtered on the home view", () => {
    expect(isInboxFiltered(home())).toBe(false);
    expect(activeInboxFilters(home())).toEqual([]);
  });

  it("COUNTS THE STATUS SEGMENT, which is the whole bug", () => {
    // The phones' Reset asked the old predicate, which excluded this, so the one
    // filter sitting closest to the Reset control was the one it could not see:
    // select Closed, press Reset, get a haptic and nothing else.
    expect(isInboxFiltered(home({ segment: "closed" }))).toBe(true);
    expect(activeInboxFilters(home({ segment: "closed" }))).toEqual(["segment"]);
  });

  it("counts Mine as the segment moving, with no separate assignee", () => {
    const mine = home({ assignedToMe: true });
    expect(isInboxFiltered(mine)).toBe(true);
    expect(activeInboxFilters(mine)).toEqual(["segment"]);
  });

  it("lets Mine subsume a named assignee rather than counting both", () => {
    // Every client sends the viewer's own id and drops the named one, and every
    // client hides the assignee control while Mine is lit. Counting it is how an
    // empty Mine tab blamed a filter with nothing to un-set.
    const both = home({ assignedToMe: true, assigneeUserId: "u-2" });
    expect(activeInboxFilters(both)).toEqual(["segment"]);
  });

  it("counts a named assignee on its own", () => {
    expect(activeInboxFilters(home({ assigneeUserId: "u-2" }))).toEqual([
      "assignee",
    ]);
  });

  it("counts each chip on its own", () => {
    for (const [key, dimension] of [
      ["tagId", "tag"],
      ["unreadOnly", "unread"],
      ["spamOnly", "spam"],
      ["snoozedOnly", "snoozed"],
      ["awaitingOnly", "awaiting"],
    ] as const) {
      const value = key === "tagId" ? "t-1" : true;
      const state = home({ [key]: value } as Partial<InboxFilterState>);
      expect(activeInboxFilters(state), key).toEqual([dimension]);
      expect(isInboxFiltered(state), key).toBe(true);
    }
  });

  it("reports every dimension at once, in the declared order", () => {
    const everything = home({
      segment: "all",
      assigneeUserId: "u-2",
      tagId: "t-1",
      unreadOnly: true,
      spamOnly: true,
      snoozedOnly: true,
      awaitingOnly: true,
    });
    expect(activeInboxFilters(everything)).toEqual([
      "segment",
      "assignee",
      "tag",
      "unread",
      "spam",
      "snoozed",
      "awaiting",
    ]);
  });
});

describe("#548 hasSecondaryInboxFilters", () => {
  it("is false when only the segment moved", () => {
    // So the empty state can use its better per-tab sentence — "No closed
    // conversations" beats "Nothing matches these filters" for somebody who has
    // selected one tab and nothing else.
    expect(hasSecondaryInboxFilters(home({ segment: "closed" }))).toBe(false);
    expect(hasSecondaryInboxFilters(home({ assignedToMe: true }))).toBe(false);
  });

  it("is true as soon as anything else is on", () => {
    expect(hasSecondaryInboxFilters(home({ unreadOnly: true }))).toBe(true);
    expect(
      hasSecondaryInboxFilters(home({ segment: "closed", tagId: "t-1" })),
    ).toBe(true);
  });

  it("is false for a named assignee that Mine has subsumed", () => {
    // The iOS empty state's exact defect: it blamed an assignee the chip strip
    // was not showing, because Mine had already absorbed it.
    expect(
      hasSecondaryInboxFilters(
        home({ assignedToMe: true, assigneeUserId: "u-2" }),
      ),
    ).toBe(false);
  });

  it("never disagrees with isInboxFiltered about the same state", () => {
    // The property that makes one list better than two predicates: secondary
    // filters existing implies something is filtered, always.
    const states = [
      home(),
      home({ segment: "closed" }),
      home({ unreadOnly: true }),
      home({ assignedToMe: true, assigneeUserId: "u-2" }),
      home({ segment: "all", tagId: "t-1", spamOnly: true }),
    ];
    for (const state of states) {
      if (hasSecondaryInboxFilters(state)) {
        expect(isInboxFiltered(state)).toBe(true);
      }
    }
  });
});
